lib/gc/Transforms/GPU/GpuToGpuOcl.cpp

```diff
@@ -0,0 +1,532 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+#include <unordered_set>
+
+#define GC_GPU_OCL_CONST_ONLY
+#include "gc/ExecutionEngine/GPURuntime/GpuOclRuntime.h"
+
+#include "mlir/Conversion/LLVMCommon/ConversionTarget.h"
+#include "mlir/Conversion/LLVMCommon/Pattern.h"
+#include "mlir/Conversion/Passes.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/GPU/Transforms/Passes.h"
+
+using namespace mlir;
+using namespace mlir::gc::gpu;
+
+namespace mlir::gc {
+#define GEN_PASS_DECL_GPUTOGPUOCL
+#define GEN_PASS_DEF_GPUTOGPUOCL
+#include "gc/Transforms/Passes.h.inc"
+} // namespace mlir::gc
+
+namespace {
+LLVM::CallOp funcCall(OpBuilder &builder, const StringRef name,
+                      const Type returnType, const ArrayRef<Type> argTypes,
+                      const Location loc, const ArrayRef<Value> arguments,
+                      bool isVarArg = false) {
+  auto module = builder.getBlock()->getParent()->getParentOfType<ModuleOp>();
+  auto function = module.lookupSymbol<LLVM::LLVMFuncOp>(name);
+  if (!function) {
+    auto type = LLVM::LLVMFunctionType::get(returnType, argTypes, isVarArg);
+    function = OpBuilder::atBlockEnd(module.getBody())
+                   .create<LLVM::LLVMFuncOp>(loc, name, type);
+  }
+  return builder.create<LLVM::CallOp>(loc, function, arguments);
+}
+
+// Assuming that the pointer to the context is passed as the last argument
+// of the current function of type memref<anyType> with zero dims. When lowering
+// to LLVM, the memref arg is replaced with 3 args of types ptr, ptr, i64.
+// Returning the first one.
+Value getCtxPtr(const OpBuilder &rewriter) {
+  auto func =
+      rewriter.getBlock()->getParent()->getParentOfType<LLVM::LLVMFuncOp>();
+  return func.getArgument(func.getNumArguments() - 3);
+}
+
+struct Helper final {
+  LLVMTypeConverter &converter;
+  Type voidType;
+  Type ptrType;
+  Type idxType;
+  mutable std::unordered_set<std::string> kernelNames;
+
+  explicit Helper(MLIRContext *ctx, LLVMTypeConverter &converter)
+      : converter(converter), voidType(LLVM::LLVMVoidType::get(ctx)),
+        ptrType(LLVM::LLVMPointerType::get(ctx)),
+        idxType(IntegerType::get(ctx, converter.getPointerBitwidth())) {}
+
+  Value idxConstant(OpBuilder &rewriter, const Location loc,
+                    size_t value) const {
+    return rewriter.create<LLVM::ConstantOp>(
+        loc, idxType,
+        rewriter.getIntegerAttr(idxType, static_cast<int64_t>(value)));
+  }
+
+  void destroyKernels(OpBuilder &rewriter, Location loc,
+                      ArrayRef<Value> kernelPtrs) const {
+    auto size = idxConstant(rewriter, loc, kernelPtrs.size());
+    auto kernelPtrsArray =
+        rewriter.create<LLVM::AllocaOp>(loc, ptrType, ptrType, size);
```

Comment:
don't we need to dealloc it after the `funcCall`?

Reply:
It's allocated on the stack - https://llvm.org/docs/LangRef.html#alloca-instruction